src/Npgsql/TypeMapping/ConnectorTypeMapper.cs

```diff
@@ -168,237 +292,242 @@ public override bool RemoveMapping(string pgTypeName)
         {
             CheckReady();
 
-            var removed = base.RemoveMapping(pgTypeName);
-            if (!removed)
+            CopyOnWriteMappings();
+
+            if (!MappingsByName.TryGetValue(pgTypeName, out var mapping))
                 return false;
 
-            // Rebind everything. We redo rather than trying to update the
-            // existing dictionaries because it's complex to remove arrays, ranges...
+            MappingsByName.Remove(pgTypeName);
+            if (mapping.NpgsqlDbType is not null)
+                MappingsByNpgsqlDbType.Remove(mapping.NpgsqlDbType.Value);
+            foreach (var clrType in mapping.ClrTypes)
+                MappingsByClrType.Remove(clrType);
+
+            // Clear all bindings. We do this rather than trying to update the existing dictionaries because it's complex to remove arrays,
+            // ranges...
             ClearBindings();
-            BindTypes();
             ChangeCounter = -1;
             return true;
         }
 
-        public override IEnumerable<NpgsqlTypeMapping> Mappings => InternalMappings.Values;
+        void CopyOnWriteMappings()
+        {
+            if (!_changedMappings)
+            {
+                // Mappings are being changed on this connector for the first time.
+                // Copy-on-write the global mappings to a mutable local Dictionary.
+                Debug.Assert(MappingsByName is IImmutableDictionary<string, NpgsqlTypeMapping>);
+
+                MappingsByName = new Dictionary<string, NpgsqlTypeMapping>(MappingsByName);
```

Comment:
I guess, the only reason we don't do it immediately is to optimize the case when the connector's type mapper is reset multiple times without changing mappings, yes? I'm not sure it's actually worth it...

Reply:
Another thing is that if the global type mappings are changed, we don't need to recopy (we just reference directly).

But looking at this again I agree, it's better to just keep this simple.